src/ALCProxy.Communication/ALCClient.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Loader;
+using System.Reflection.Emit;
+using System.IO;
+
+namespace ALCProxy.Communication
+{
+    internal delegate object ServerCall(MethodInfo info, IList<object> args, IList<Type> types);
+    /// <summary>
+    /// This currently is designed to only work in-process
+    /// TODO: set up to allow for construction of out-of-proc proxies
+    /// </summary>
+    public abstract class ALCClient : IProxyClient
+    {
+        //Can't make this an IServerObject directly due to the type-loading barrier
+        protected object _server;
+        protected string _serverTypeName;
+        protected Type _intType;
+        internal ServerCall _serverDelegate;
+        protected Type _serverType;
+
+        private StackTrace _stackTrace;
+        public ALCClient(Type interfaceType, string serverName, Type serverType)
+        {
+            if (interfaceType == null || serverName == null)
+                throw new ArgumentNullException();
+            _serverType = serverType;
+            _intType = interfaceType;
+            _serverTypeName = serverName;
+#if DEBUG
+            _stackTrace = new StackTrace(true); //holds information for debugging purposes
+#endif
+        }
+        private Type FindTypeInAssembly(string typeName, Assembly a)
+        {
+            //find the type we're looking for
+            Type t = a.GetType(typeName);
+            if (t == null)
+            {
+                t = a.GetType(a.GetName().Name + "." + typeName);
+                if (t == null)
+                {
+                    throw new TypeLoadException("Proxy creation exception: No valid type while searching for the given type: " + typeName + " || " + a.GetName().Name + "." + typeName);
+                }
+            }
+            return t;
+        }
+        /// <summary>
+        /// Creates the link between the client and the server, while also passing in all the information to the server for setup
+        /// </summary>
+        /// <param name="alc">The target AssemblyLoadContext</param>
+        /// <param name="typeName">Name of the proxied type</param>
+        /// <param name="assemblyName">path of the assembly to the type</param>
+        /// <param name="genericTypes">any generics that we need the proxy to work with</param>
```

Comment:
If the plan is to get public APIs documented with xml comments, make sure to add exception tags as well to document what exceptions are thrown and when.